### PR TITLE
new connection spec

### PR DESCRIPTION
**Pull request by adityathebe**
resolves: https://github.com/flanksource/duty/issues/629

### CODE REVIEW

File: api/v1/connection_types.go [hunk 1]
```diff
@@ -5,16 +5,188 @@ import (
...
+
+type ConnectionHTTP struct {
+	URL         types.EnvVar  `json:"url"`
+	InsecureTLS bool          `json:"insecureTLS,omitempty"`
+	Username    *types.EnvVar `json:"username,omitempty"`
+	Password    *types.EnvVar `json:"password,omitempty"`
+}
+
+type ConnectionSFTP struct {
+	Host     types.EnvVar `json:"host"`
```
Comment (moshloop) [inline api/v1/connection_types.go:143]: This and the other folder types should have a `path`

File: api/v1/connection_types.go [hunk 2]
```diff
@@ -5,16 +5,188 @@ import (
...
+	Properties types.JSONStringMap `json:"properties,omitempty"`
+
+	AWS   *ConnectionAWS   `json:"aws,omitempty"`
+	Azure *ConnectionAzure `json:"azure,omitempty"`
+	GCP   *ConnectionGCP   `json:"gcp,omitempty"`
+
+	Folder     *ConnectionFolder     `json:"folder,omitempty"`
+	Git        *ConnectionGit        `json:"git,omitempty"`
+	GitHub     *ConnectionGitHub     `json:"github,omitempty"`
+	Postgres   *ConnectionPostgres   `json:"postgres,omitempty"`
```
Comment (moshloop) [inline api/v1/connection_types.go:163]: I think we need mysql/mssql/s3/gcs/azureDevops

File: api/v1/connection_types.go [hunk 3]
```diff
@@ -5,16 +5,232 @@ import (
...
+	// List of recipient e-mails
+	ToAddresses []string `json:"toAddresses"`
+
+	// The subject of the sent mail
+	Subject string `json:"subject"`
+
+	// Auth - SMTP authentication method
+	// Possible values: None, Plain, CRAMMD5, Unknown, OAuth2
+	Auth string `json:"auth"`
+}
```
Comment (moshloop) [inline api/v1/connection_types.go:66]:
```suggestion
type ConnectionSMTP struct {
	Host     string       `json:"host"`
	Username types.EnvVar `json:"username,omitempty"`
	Password types.EnvVar `json:"password,omitempty"`

	// Default: false
	InsecureTLS bool `json:"insecureTLS,omitempty"`

	// Encryption Method
	// 	Defulat: auto
	// 	Possible values: None, ExplicitTLS, ImplicitTLS, Auto
	Encryption string `json:"encryption,omitempty"`

	// SMTP server port
	// 	Default: 587
	Port int `json:"port,omitempty"`

	// Email address that the mail are sent from
	FromAddress string `json:"fromAddress"`

	// List of recipient e-mails
	ToAddresses []string `json:"toAddresses"`

	// The subject of the sent mail
	Subject string `json:"subject"`

	// Auth - SMTP authentication method
	// Possible values: None, Plain, CRAMMD5, Unknown, OAuth2
	Auth string `json:"auth,omitempty"`
}
```

File: api/v1/connection_types.go [hunk 4]
```diff
@@ -5,16 +5,232 @@ import (
...
+	Bucket        string `json:"bucket"`
+}
+
+type ConnectionAWS struct {
+	Region      string       `json:"region"`
+	Profile     string       `json:"profile"`
+	InsecureTLS bool         `json:"insecureTLS,omitempty"`
+	AccessKey   types.EnvVar `json:"accessKey"`
+	SecretKey   types.EnvVar `json:"secretKey"`
+}
```
Comment (moshloop) [inline api/v1/connection_types.go:116]:
```suggestion
type ConnectionAWS struct {
	Region      string       `json:"region,omitempty"`
	Profile     string       `json:"profile,omitempty"`
	InsecureTLS bool         `json:"insecureTLS,omitempty"`
	AccessKey   types.EnvVar `json:"accessKey,omitempty"`
	SecretKey   types.EnvVar `json:"secretKey,omitempty"`
}
```

File: api/v1/connection_types.go [hunk 5]
```diff
@@ -5,16 +5,232 @@ import (
...
+
+type ConnectionAzureDevops struct {
+	Organization        string       `json:"organization"`
+	PersonalAccessToken types.EnvVar `json:"personalAccessToken"`
+}
+
+type ConnectionGCP struct {
+	Endpoint    types.EnvVar `json:"endpoint"`
+	Certificate types.EnvVar `json:"certificate"`
```
Comment (moshloop) [inline api/v1/connection_types.go:131]:
```suggestion
	Endpoint    types.EnvVar `json:"endpoint,omitempty"`
	Certificate types.EnvVar `json:"certificate,omitempty"`
```

File: api/v1/connection_types.go [hunk 6]
```diff
@@ -5,16 +5,232 @@ import (
...
+	Region      string       `json:"region"`
+	Profile     string       `json:"profile"`
+	InsecureTLS bool         `json:"insecureTLS,omitempty"`
+	AccessKey   types.EnvVar `json:"accessKey"`
+	SecretKey   types.EnvVar `json:"secretKey"`
+}
+
+type ConnectionAzure struct {
+	ClientID     types.EnvVar `json:"clientID"`
+	ClientSecret types.EnvVar `json:"clientSecret"`
```
Comment (moshloop) [inline api/v1/connection_types.go:120]:
```suggestion
	ClientSecret types.EnvVar `json:"clientSecret,omitempty"`
```

File: api/v1/connection_types.go [hunk 7]
```diff
@@ -5,16 +5,232 @@ import (
...
+}
+
+type ConnectionPostgres struct {
+	Host     types.EnvVar `json:"host"`
+	Username types.EnvVar `json:"username"`
+	Password types.EnvVar `json:"password"`
+}
+
+type ConnectionMySQL struct {
+	Host     types.EnvVar `json:"host"`
```
Comment (moshloop) [inline api/v1/connection_types.go:87]: All the SQL connections need to support specify a host + database and it builds the URL, or specifying just the URL

File: api/v1/connection_types.go (hunk 3 above)
Comment (adityathebe) [reply api/v1/connection_types.go:66]: Should have been optional

**Comment by adityathebe** [comment]
merge after: https://github.com/flanksource/duty/pull/641

File: api/v1/connection_types.go [hunk 8]
```diff
@@ -5,16 +5,256 @@ import (
...
+	PersonalAccessToken types.EnvVar `json:"personalAccessToken"`
+}
+
+type ConnectionGit struct {
+	URL         string        `json:"string"`
+	Ref         string        `json:"ref"`
+	Certificate *types.EnvVar `json:"certificate"`
+	Username    *types.EnvVar `json:"username"`
+	Password    *types.EnvVar `json:"password"`
+	SSH         *types.EnvVar `json:"ssh"`
```
Comment (moshloop) [inline api/v1/connection_types.go:176]: Why do we have both certificate and SSH? the auth is also not required

File: db/connections.go [hunk 9]
```diff
@@ -27,6 +31,268 @@ func PersistConnectionFromCRD(ctx context.Context, obj *v1.Connection) error {
...
+		}
+	}
+
+	if obj.Spec.MSSQL != nil {
+		dbObj.Type = models.ConnectionTypeSQLServer
+
+		if obj.Spec.MSSQL.URL.String() != "" {
+			dbObj.URL = obj.Spec.MSSQL.URL.String()
+		} else {
+			dbObj.URL = "Server=$(properties.host);Database=$(properties.database);User Id=$(username);Password=$(password);Encrypt=$(properties.insecure_tls)"
```
Comment (moshloop) [inline db/connections.go:151]: I don't think encrypt is the right mapping for insecure tls

**Pull request merged**